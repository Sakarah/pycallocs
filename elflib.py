import importlib
import importlib.abc
import importlib.machinery
import sys
# Re-export everything from the C extension module
from allocs import *

# These can be modified by the user to change the finder search paths
lib_extension = ".so"

# The blank search path has a special meaning as dlopen is using its own search 
# paths when the file name does not contain any /
__path__ = ["./", ""]

class LibraryFinder(importlib.abc.MetaPathFinder):
    """
        Meta path finder searching foreign libraries usable
        with allocs.ForeignLibraryLoader.
    """

    def find_spec(fullname, path, target):
        if path is None:
            # We are defining __path__ so it should never be None
            return None

        # We only try to load modules below ourselves
        modpath = fullname.split(".")
        if len(modpath) != 2 or modpath[0] != __name__:
            return None
        name = modpath[1]

        for base_path in path:
            try:
                filename = base_path + name + lib_extension
                loader = LibraryLoader(filename)
                return importlib.machinery.ModuleSpec(fullname, loader, origin=filename)
            except ImportError:
                continue

        return None

# We must register ourselves before the default meta path finder to prevent exception
# ImportError: dynamic module does not define module export function (PyInit_%)
sys.meta_path.insert(0, LibraryFinder)
