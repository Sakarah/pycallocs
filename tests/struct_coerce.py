import elflib
elflib.__path__.append("libs/")
from elflib import composite as m

class Hw:
    def __init__(self, h, w):
        self.hello = h
        self.world = w

class Hcw(Hw):
    def __init__(self, h, c, w):
        Hw.__init__(self, h, w)
        self.crazy = c

a = Hw(42, 9.99)
m.print_hw(a)
b = Hcw(-1, "foo", float('nan'))
m.print_hw(b)
m.print_hw((49,3))
m.print_hw({'hello':0, 'world':2.71828})
m.print_hw({'hello':1, 'crazy':None, 'world':3.14159})

try:
    m.print_hw((a, b))
    exit(1)
except TypeError:
    pass

try:
    m.print_hw((-1, "foo", float('nan')))
    exit(1)
except TypeError:
    pass
