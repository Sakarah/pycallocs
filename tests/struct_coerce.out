Hello: 42, World: 9.99
Hello: -1, World: nan
Hello: 49, World: 3.00
Hello: 0, World: 2.72
Hello: 1, World: 3.14
